app/lib/package/backend.dart

```diff
@@ -891,28 +893,48 @@ class PackageBackend {
 
   /// Finishes the upload of a package and returns the list of messages
   /// related to the publishing.
-  Future<List<String>> publishUploadedBlob(String guid) async {
+  Future<List<String>> publishUploadedBlob(String uploadGuid) async {
     final restriction = await getUploadRestrictionStatus();
     if (restriction == UploadRestrictionStatus.noUploads) {
       throw PackageRejectedException.uploadRestricted();
     }
     final agent = await requireAuthenticatedClient();
-    _logger.info('Finishing async upload (uuid: $guid)');
+    _logger.info('Finishing async upload (uuid: $uploadGuid)');
     _logger.info('Reading tarball from cloud storage.');
 
     return await withTempDirectory((Directory dir) async {
-      final filename = '${dir.absolute.path}/tarball.tar.gz';
-      final info = await _incomingBucket.tryInfo(tmpObjectName(guid));
+      // Check the existence of the uploaded file
+      final uploadObjectName = tmpObjectName(uploadGuid);
+      final info = await _incomingBucket.tryInfo(uploadObjectName);
       if (info?.length == null) {
         throw PackageRejectedException.archiveEmpty();
       }
+
+      // Create a temporary copy that we will continue working with.
+      // This will protect us against the unlikely scenario where the
+      // uploaded blob changes during this processing.
+      final workGuid = createUuid();
+      final workObjectName = '${tmpObjectName(workGuid)}-$uploadGuid';
+      try {
+        await _storage.copyObjectWithRetry(
+          _incomingBucket.absoluteObjectName(uploadObjectName),
+          _incomingBucket.absoluteObjectName(workObjectName),
+        );
+      } catch (e) {
+        throw InvalidInputException(
+            'Failed to copy uploaded file (uuid:$uploadGuid).');
```

Comment:
Perhaps add the error 
```suggestion
            'Failed to copy uploaded file (uuid:$uploadGuid): $e.');
```

Reply:
Hm, it may be worth logging, but not exposing to the end user.

Reply:
Ah yes, sorry, the user should definitely **not** get this error 😅.